zilliqa/src/consensus.rs

```diff
@@ -635,6 +661,25 @@ impl Consensus {
             );
 
             let mut block_receipts: Vec<TransactionReceipt> = Vec::new();
+
+            if !transactions.is_empty() {
+                trace!("applying {} transactions to state", transactions.len());
+            }
+
+            // Must make sure state root hash is set to the parent's state root hash before applying transactions
+            if self.state.root_hash()? != parent.state_root_hash() {
+                warn!("state root hash prior to block execution mismatch, expected: {:?}, actual: {:?}", parent.state_root_hash(), self.state.root_hash()?);
+            }
+
+            if head_block.hash() != parent.hash() || block.number() != head_block.header.number + 1
```

Comment:
No, the parent hash mismatch is not good enough because if there are no transactions in the network, both branches will have the same hash for every block. Part of dealing with the fork is making sure you re-point head_block, which I have written it such that head_block is the absolute reference point for the chain (finalized block being more of a side detail).

WRT should the check be `block.number() < head_block.header.number` - I see where you are getting at but I don't think that `deal_with_fork` should be able to unwind to genesis. Consider the loop:

```
        while self.head_block().header.number >= block.header.number {
```

It only loops while head block is >= proposal. So the final head_block height should just be proposed block - 1 if it is working correctly.

As for the case where the block height is >> head_block - I believe that this should be impossible in a healthy system, in any case there is a check for it in `check_block`: 

```
        // Blocks must be in sequential order
        if block.header.number != parent.header.number + 1 {
```

Perhaps it can occur while you are syncronizing and you approach HEAD.